### ISSUE TITLE
Beta: Package Editor saves wrong class

### ISSUE THREAD

**Issue by Mgamerz** [ME3Explorer/ME3Explorer#368]
I am trying to figure out some issues in my ME3 GUI Transplanter tool. As I was trying to figure out the cause I have been trying to replicate its functionality in ME3Explorer itself, but I have found an odd bug. I open up BioD_CitHub_Underbelly (from Backoff mod if that matters) in Package Editor, and I open the GUI_SF_ME2_AreaMap.ME2_AreaMap export. It takes a long time to load because the array is binary data with hundreds of thousands of items (This should really not be parsed unless specifically done so by the user, as it makes the program super slow, and most users won't want to modify binary arrays). 

In interpreter, I modify GFX to SWF, the first bytes of the binary, which is a meaningless change to the game engine. I save the file using the normal save method. The class in ME3EXP still shows GfxMovieInfo. When I ran my ME3 GUI tool over it again to see if the saving worked, it shows no GfxMovieInfo exports found - so I reload the file in Me3Explorer, and the class of that export is now something else (2PolySquarePlane_BottomAxis). I did not change anything beyond 3 bytes in the binary data section - this change should not have occurred. 

I tested this in both 2.0.3 and 2.0.4 and was able to reproduce it in both.

EDIT: It seems that saving the imports list for these builds is just saving the same exact thing for every export, like an i++; line of code wasn't added somewhere in a loop or something. The original import lists is a bunch of different ones, once saved it is a list of literally all the same.

This does not appear to be bugged on reconstructive save.

**Comment by SirCxyrtyx** [comment]
You talk about Reconstruction save and import list saving in Package Editor. Those are things I added post 2.0.4 and haven't been put into a release yet. Are you using a version you built from the current state of the Beta branch?

> It seems that saving the imports list for these builds is just saving the same exact thing for every export, like an i++; line of code wasn't added somewhere in a loop or something. The original import lists is a bunch of different ones, once saved it is a list of literally all the same.

Weird, I used a ForEach loop, which is almost the exact same code I used in the reconstruction save. Not in a position to investigate further at the moment, as I am away from my coding computer until the 22nd.

**Comment by Mgamerz** [comment]
My gui transplanter tool has some older version of pccobject.cs that is
having trouble saving pcc's where I update an array property (properly,
updating size, and appending other info). In an attempt to see if I could
fix it I took the latest copy from ME3Explorer github, which apparently is newer
than 2.0.4, and noticed the reconstructive save. I tried that, and it seems
to work. Now, that won't work on dlc pccs, so I also tested altsave/normal
save and that still is broken. So I went into 2.0.3 of ME3Explorer and
manually test-modified this PCC to see if I could get ME3Explorer to have
the same issue, but instead I found that the imports list is saving wrong,
somehow, on the standard "Save" button (didn't try saving (old)). This is 2.0.3/2.0.4 public releases, nothing I built myself.

TLDR: My transplanter tool has the not-yet-finished PCCobject (I'll look at
2.0.4) which seems to have a working reconstructive save, but the
pccobject.cs in 2.0.3/2.0.4 seems to save the imports list wrong. I looked
over the saving code vs old ones, and I don't really see anything wrong.
I'll see if I can dig something up.

(Reconstructive save seems OK - the main save used in Package Editor doesn't seem to be)

Note I am not asking for fixes for the file in my transplanter tool, but since it uses the same exact code as me3explorer, the bug probably exists there too, but not exactly sure what's triggering it right now. Not asking you to fix that bug, but the imports bug applies to both.

**Comment by Mgamerz** [comment]
There seems to be some changes when you switched interpreter in for hex editor in this commit:

https://github.com/ME3Explorer/ME3Explorer/commit/5397321295b0501370491ada057d426c14f92618#diff-f0f204ad0c525e7b6ede70f3fd0015b3

I'll make a few changes and see if I can fix it and then tell you what I did to fix it so you can merge it with your 2.0.5 code. If I find any other bugs (I might see a couple here) I'll include those as well.

**Comment by Kinkojiro** [comment]
I have been noticing a bug I cannot track down that changes the size of certain files just fractionally between saves, even when I have only changed a property's value (i.e. not added or removed anything).  Could it be the same thing?

**Comment by Mgamerz** [comment]
Could be related. The only thing I did was a hex edit, nothing else. My file didn't change size, but it appears there is some garbage/unused data from the old PCC before giftfish modified it and me3explorer is unhappy about it for some reason.

I'll be looking into this tonight, though I don't k ow enough about your issue to look specifically for that one (interpreter has changed a lot these days)

**Comment by Mgamerz** [comment]
Additionally - the reason I am aware of this bug is that when I ran my transplanter tool from my public vs ide build - on the same input source - I got different results. As you said kinko, mine were like 2KB difference, but there should have been no difference as the task was deterministic.

**Comment by Mgamerz** [comment]
I've done some more testing. There is a bug in the saving code of PCCs somewhere in 2.0.4 - I just tested altsave after hex editing an export (ascii GFX to SWF, nothing else, in file BioD_Nor #556) and once saved the pcc no longer loads. Throws an arithmetic error as it is reading past the end of the file, meaning its reading the incorrect data somewhere (or writing wrong somewhere) This is the same issue I am facing in my transplanter code. Strangely, I'm having some issues reproducing it again...

Given the complexity of pcc saving, this may take some time to trace... Saved the file with just standard save and it seemed to save OK, no class name changes either.

**Comment by SirCxyrtyx** [comment]
The reason I wrote the reconstruction save method is that I was having the same problems with PckEd altsave, where it would produce unreadable files. Rather than try to fix the unnecessarily complicated code, I just made a new save method. I intend to just remove the altsave entirely.

**Comment by Mgamerz** [comment]
I think I'm making some progress on the altsave/savetofile issue. I am looking at some files from expanded galaxy mod, and the name table is at the normal position (~0x90), but there's a full duplicate name table at the bottom, with what looks like kinko's edits. Part of the pcc code doesn't seem to account for the name table at this position, because it is calculating the place to save new/changed exports as the area directly after the current export data (lastexportoffset). This conflicts with placing the name table at the end. As such, when I am changing this SWF export, it is overwriting the name table that has been appended to the bottom of this pcc file. When the pcc file loads again, it reads my SWF data as the name table and blows up. This would explain why I am having a hard time reproducing it, as it seems files were saved with whatever method the last tool to edit it used. This would also explain why the file size isn't changing, because the name table at the end is considered "garbage data" because theres nothing properly referencing it.

Still trying to find that class changing bug, but it's not showing up as reliably as this morning.

I've been using your reconstructive save, the only bug I have found so far is that it does not update the compression flag. So it will try to decompress a file that's already decompressed and it won't be happy. I fixed it by just setting this.bCompressed = false when saving.

Edit: yep, here in saveToFile I am seeing this issue. Here's a pic of exactly whats happening.
![image](https://cloud.githubusercontent.com/assets/2738836/15348180/98f02dec-1c85-11e6-8b08-64374c792c6a.png)

The name table is being overwritten. I'm amazed this issue hasn't come up more. This probably has to do with how each type of save appends data. One could probably add new export data after the name table, and on each new save, just append new export data and then append the new nametable and adjust the headers accordingly. One could account for the size of the nametable from the last data offset.

**Comment by Mgamerz** [comment]
Updated with picture showing the issue in debugger. Working on a fix. Other issues are things like the header size being computed as the size from 0 to the nametable, when the nametable is at the end, is causing an issue.

I'm looking further into the code, and @Kinkojiro I have noticed that the imports list is always being saved after the name table. If you added names to your pcc, the name table is at the end, so it is writing them to the end of the file regardless of you editing it. So it is appending even if there are no changes. This may be why your file size is changing. In fact its appending just about everything except actual unchanged export data it seems.

Edit: Seems like I got basic save to work (so far on a single file). This is good news :100: I'll do some more testing tomorrow on more files. I have a verifypcc method in my gui transplanter that essentially just forces lots of random data to load and make sure no exceptions occur, and it passed.

The bug I mentioned in the OP of this thread may be related, since the imports table may have been overwritten by a big chunk of same data, resulting in a class change since the import entry was actually changed. I commited my fixed PCCObject to my GUI Transplanter, and I'll insert it into me3explorer and see if I can force it to retrigger bug. If you want a peek at how I fixed it here's the commit https://github.com/Mgamerz/ME3-GUI-Transplanter/commit/8965dcedf5dadcd3e15624305c3f065fd051b88e

Essentially I just checked if currnet writing offset = names/import offset, and if it is, I appended the size of the nametable/import table before writing the new export data.

**Comment by Kinkojiro** [comment]
In my experience it doesn't overwrite parts of the table (which would cause the file to be shredded) but it does change the size so requiring an autoTOC.   It is random though, it doesn't seem to happen regularly which makes it very hard to pin down.

**Comment by Mgamerz** [comment]
Try cases of only adding a single name, or editing a file that has imports changed or names edited in the previous save.

**Comment by Mgamerz** [comment]
I am able to reproduce this, unreliably, via adding a name in package editor and doing save and then doing it again with alt save.

Also the 'alt save' in the UI is 'save' in code and save in UI is altsave in code. I applied the fix I used for GUI Transplanter (which works A OK) but it doesn't seem to be working here. In my tool I always ave decompressed though, in this tool it always saves compressed (I opened a decompressed file and it saved compressed)

**Comment by SirCxyrtyx** [comment]
So, let me see if I've got this straight. The bug is only in the PCCObject::saveToFile function? That is being removed in 2.0.5

Good catch with setting bCompressed to false in the reconstruction save.

**Comment by Mgamerz** [comment]
Save to file also exists in kfreonlib correct?

**Comment by SirCxyrtyx** [comment]
Oh, ugh. It does indeed. Only thing that should be using that is TPF Tools, Texplorer, and Modmaker though. And with what those tools do I would assume they'd be using the altSaveToFile function. I can do a find all references check in VS to make sure. But even if they are using saveToFile, this bug never seems to have cropped up in those tools, and I'm not touching them anyways.

That does remind me that the ME1 and ME2 versions will have saveToFile as well. I'll have to remove those too and make reconstruction saves for them instead.

**Comment by Mgamerz** [comment]
I looked through your code on those two tools, and it doesn't seem to save
the same way, it saves in a linear fashion and sets offset as it goes,
which is how it is supposed to work. I think WV tried to optimize saving
which is where the bug came from.
